### DISCUSSION TITLE
INTERNAL Error: Comparison on NULL values while building Parquet files

### DISCUSSION THREAD

**Discussion by parvizfarnia** [duckdb/duckdb#6304] [Q&A]
Hello everyone,

I'm using the most recent version 0.7.0 of DuckDB to be able to create Parquet files via Hive-style (key1=value/key2=value, ...)

I checked the syntaxe on the following page https://github.com/duckdb/duckdb/pull/5964 and I put the following in my Python program:

`db_connection.execute("copy mytab to 'D:\\tmp\\parquet_data_dir' (format parquet, partition_by(col1, col2, col3, col4)))`

After a few minutes this gave me the following error message: 

```
line 228, in main
    db_connection.execute(
duckdb.InternalException: INTERNAL Error: Comparison on NULL values
```

I import a CSV file to DuckDB and then I want to export it as a multi-file dataset. I didn't check whether there was any NULL values on the columns in my CSV because there was no business restriction in this regard. I also checked the online documentation of Apache Parquet at chapter [Nulls](https://parquet.apache.org/docs/file-format/nulls/). I didn't find anything stating that NULL values were forbidden in Parquet.

What am I doing wrong?

Thanks in advance

**Comment by Mytherin** [comment]
Thanks for the report! This seems like a bug in DuckDB. As a rule any `InternalException` is a bug. Could you file a reproducible bug report?

**Reply by parvizfarnia** [reply]
Thank you for confirming this. Sorry, I am very beginner in both DuckDB and Github. When you say "reproducible bug report" you mean that I have to create a new thread in [Issues](https://github.com/duckdb/duckdb/issues) with my program and a sample CSV file?

**Reply by Mytherin** [reply]
Yes, exactly

**Reply by parvizfarnia** [reply]
Yes, of course. But first, I think I have to investigate more in my data file. I was actually preparing a reproducible test case with a sample CSV file in which I intentionally put several NULL values (empty values in CSV) on multiple columns. However, this time everything worked like a charm and the entire hierarchical structure (hive-style) was built without any problem. 

However, based on what you wrote this should be rather a bug (`InternalException`). The problem is that the error message: 

`duckdb.InternalException: INTERNAL Error: Comparison on NULL values` 

doesn't provide enough details about the exact location (or at least on which column(s)) where the NULL value was located or to find somewhat what portion of the file was problematic. And my CSV is quite big (more than 30 000 000 rows). 

I try to investigate and see how can I reproduce it with a simple test case.

**Reply by parvizfarnia** [reply]
I am still trying to produce a pertinent test case. Now I have even a new error message when I try to make a parquet (with partition) on a big CSV file (over 30 000 000 rows) :

`duckdb.OutOfMemoryException: Out of Memory Error: failed to pin block of size 262144`

I found among the already declared issues a thread with the same error message : https://github.com/duckdb/duckdb/issues/2907 But I don't know whether they are related.

**Reply by parvizfarnia** [reply]
Good news! I think I finally found the cause of the original problem (based on multiple tests that I have done so far). If in my program I write :

`db_connection.execute("copy mytab to 'D:\\tmp\\parquet_data_dir' (format parquet, partition_by(col1, col2, col3, col4)))`

And the first column (so `col1` in the above example) in `partition_by` clause is empty (NULL), then I get the error message : 

`duckdb.InternalException: INTERNAL Error: Comparison on NULL values
`

The fact that any of other columns `col2, col3 or col4` is NULL seems to have no such impact. Therefore, only the first column in `partition_by`, if it is NULL, appartently is problematic.

So I'm going to create a new issue in a few minutes with a simple and reproducible test case.

However for the other error message (`duckdb.OutOfMemoryException: Out of Memory Error: failed to pin block of size 262144`) I have no clue.

**Reply by Mytherin** [reply]
Thanks for the extra info - I have managed to reproduce it and pushed a fix (#6336). The high memory usage is expected as the partition by currently materializes a lot of data in memory first. Try specifying a temporary directory so the system can offload the data to disk (`SET temp_directory='my_temp_dir.tmp'`).

**Reply by parvizfarnia** [reply]
Thanks a lot.

And here is the test case that I provided : https://github.com/duckdb/duckdb-web/issues/556 

I hope that might be useful

**Reply by parvizfarnia** [reply]
When you say "I pushed a fix", does this mean that if I do "`pip install duckdb==0.7.0`" once again, I'll be able to obtain a new version including the fix you have provided?

**Reply by Mytherin** [reply]
No, it means the fix was pushed to the development branch. It will be released in v0.7.1 (which should be released in the coming weeks).

**Reply by parvizfarnia** [reply]
Ok, thank you for confirmation. Meanwhile, I'm going to apply your suggested solution about the memory

`SET temp_directory='my_temp_dir.tmp'`

And I will let you know whether it solves my problem.

Thank you very much for your time and your help.

**Reply by parvizfarnia** [reply]
I run the program once again by adding the following to my program : 

`SET temp_directory='my_temp_dir.tmp'`

And here is the description of our datalab environment in which I run the program without even setting any max_size memory available for DuckDB: 

```
CSV file size to import :  ~3GB (almost 33 000 000 rows)
Available RAM memory in the datalab : 360 GB
Available disk : 300 GB
```

And this time, after running for about two hours, the program stopped and printed the following error message: 

`duckdb.OutOfMemoryException: Out of Memory Error: could not allocate block of 262144 bytes (268435225212/268435456000 used)
`

During the execution time, I was checking regularly the RAM in our shared/multi-user data lab environment. I can confirm that not even 20% of the available RAM was being used. So the excessive number 268435225212 bytes, (that is, 268 GB!!) memory usage is completely a mystery for me and I don't see how the program didn't get the required memory, not to mention that hundreds of GB were available on the disk for the TMP directory.

**Reply by parvizfarnia** [reply]
Do you put any limit on the created TMP directory in your code or so long as there is space on the disk, if required by the program, the temporary directory can go automatically as big as needed? 

Here is how I defined my TMP directory in my program:

```
db_connection = duckdb.connect(
     database='...', 
     read_only=False
)

db_connection.execute("set temp_directory='my_temp_dir.tmp'")
```

While the program was running, I saw two files were created in the TMP directory which I had specified n my Python program, one with almost 1.2 GB and the other with almost 200 MB. The whole CSV file (uncompressed) was about 3GB. I'm trying to understand what went wrong. I can hardly believe that nobody has ever managed to parition a parquet based on multiple columns on a file which is just 3GB. Many work on files with billions of rows.
